src/language/ast.js

```diff
@@ -538,3 +538,56 @@ export type DirectiveDefinitionNode = {
   +arguments?: $ReadOnlyArray<InputValueDefinitionNode>,
   +locations: $ReadOnlyArray<NameNode>,
 };
+
+export type KindToASTNodeType = {|
```

Comment:
It would be nicer to instead use `$PropertyType<T, 'kind'>` instead of manually assembling this type

Reply:
My goal was to provide correct types for `node` argument of callback depending on how it registered, e.g.:
```ts
visit(ast, {
  Document(node) {
     // node should be DocumentNode and not ASTNode
  }
});
```
The only way to do this is to create the type like that:
```flow
import type {
  DocumentNode,
  // ...
} from './ast';

type Visitor = {
   enter: VisitorFn<ASTNode> | { 
      Document: VisitorFn<DocumentNode>,
      ...
   },
   Document: VisitorFn<DocumentNode> | { enter?: VisitorFn<DocumentNode>, leave?: VisitorFn<DocumentNode>}
  // ...
}
```
So I need to list every node type three times. That's why I tried to reduce it to the following:
```flow
import type { KindToASTNodeType } from './ast';

type KindVisitor = {|
  ...$ObjMap<KindToASTNodeType, <T>(T) => VisitorFn<T>>,
|};

type Visitor = {|
  enter?: VisitorFn<ASTNode> | KindVisitor,
  leave?: VisitorFn<ASTNode> | KindVisitor,

  ...$ObjMapi<KindToASTNodeType , <T>(T) => NodeVisitor<T>>,
|};
```
I really like to properly type `visitor` argument because without it you can achieve type-safety inside visitor callbacks. 
I'm new to flow, so can you please  give example how I can achieve the same result with `$PropertyType<T, 'kind'>`? 
Or should I just list every node type without trying to use clever tricks?